### PR TITLE
Replicator

### PR DESCRIPTION
**Pull request by kore**
Documented workings of a replicator to synchronize the content repository with external systems (Search).

### CODE REVIEW

**Comment by andrerom** [comment]
As this is a psoposal, and not something we have implemented, it can not be in the main specification folder.

In legacy we had versioned specifications, ref `doc/specifications/trunk`, however I don't think I ever liked that structure as it made it hard to find stuff. So maybe just `doc/proposed/specifications`, what do you think @bdunogier?

**Comment by bdunogier** [comment]
`doc/specifications/proposed` would make more sense, but it would force us to move the existing ones.  `doc/specifications/implemented` could make sense though.

**Comment by bdunogier** [comment]
I'd suggest you use the `.rst`extension to makes the file renderable by github.

**Comment by andrerom** [comment]
> doc/specifications/proposed would make more sense, but it would force us to move the existing ones. doc/specifications/implemented could make sense though.
> I'd suggest you use the .rstextension to makes the file renderable by github.

+1 on that structure.

**Comment by kore** [comment]
Moved and renamed as requested. Rebased on master.

**Comment by andrerom** [comment]
As this is spec proposal, I'm +1 on it in current form, further iterations to improve the spec can be done later, or expanding it with design doc one we get around to prioritize this.

@bdunogier you're also ok with this as a convention for proposals from now on?

**Comment by bdunogier** [comment]
Okay with the structure, yes. 

We should write some guidelines about it soon, so that we have a common understand of what it is meant for, what should be added there, and what the process is like.

File: doc/specifications/proposed/replicator/architecture.rst [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+multiple systems, so **do not** rely on that.
+
+A simple auto-increment in the write-master might be the simplest solution. You
+should take care for possible overflows, though. The revision might increase
+quite fast in scenarios with a lot of updates.
+
+Idempotent Updates
+------------------
+
+In some cases it is helpful if all updates are idempotent, thus applying the
```
Comment (gggeek) [inline doc/specifications/proposed/replicator/architecture.rst]: Tip: ALL updates MUST be idempotent. This makes the protocol much more stable, reliable, easy to implement.

Ideally you would have a log somewhere of the unexpected situations (eg. master has deleted a content but that content is not there on the slave), and design your apis so that those warnings could easily be turned into blockers if needed.

File: doc/specifications/proposed/replicator/architecture.rst [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+Compaction
+----------
+
+One problem with this approach is that if the replication target had gone away
+every revision of every document will be re-transmitted in order. This can lead
+to quite long initializations.
+
+An option commonly used is to compact the database, thus removing deleted
+objects and older revisions so that they won't be re-transmitted. One must pay
+attention to not affect data integrity when doing this.
```
Comment (gggeek) [inline doc/specifications/proposed/replicator/architecture.rst]: The big problem with legacy content engine is that a lot of content operations are not atomic.
Which makes compaction an hazard.
E.g. add content, edit content, delete content => compaction means the 3 events could be removed from the queue.
But what if any of the slaves has already synchronized the 1st event?

File: doc/specifications/proposed/replicator/architecture.rst [hunk 3]
```diff
@@ -0,0 +1,142 @@
+==========
+Replicator
+==========
+
+This document describes a common synchronization mechanism between two systems,
+which can cope with write failures, network downtimes and other types of
+partitions. The approach only works with a single write master (Single Source
+of Truth) and any number of read slaves.
```
Comment (gggeek) [inline doc/specifications/proposed/replicator/architecture.rst]: TIP: design the system/api to be resilient to edits done on the slave as well. Maybe not as a guarantee that system will work, but this will happen in real life

File: doc/specifications/proposed/replicator/architecture.rst [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+It generally can be described as a three step process:
+
+1) Fetch last revision from target
+
+   The target is asked for the last revision the target knows. This is the
+   revision of the last successful write operation. For the first request the
+   target may return something like ``null``.
+
+2) Fetch all changes since that last revision from source
```
Comment (gggeek) [inline doc/specifications/proposed/replicator/architecture.rst]: NB: this means that master should store more or less indefinitely all data that it normally discards. This might pose a storage problem.

Eg. version 1-100 of object X, which is then deleted at revision 101 have to be kept, most likely in a separate storage space, until all slaves have duplicated them.

In the legacy content engine a lot of operations are permitted which allow the editor to destroy historical content, eg. remove old revisions, remove languages (which alters old revisions), remove contents.

File: doc/specifications/proposed/replicator/architecture.rst [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+General Workings
+================
+
+The general scheme for synchronizations is fairly simple:
+
+.. image:: synchronize.png
+
+It generally can be described as a three step process:
+
+1) Fetch last revision from target
```
Comment (gggeek) [inline doc/specifications/proposed/replicator/architecture.rst]: The most common usecase would be to seed the target using a copy of the master database. I think this should be mentioned in the spec

File: doc/specifications/proposed/replicator/architecture.rst [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+   **must** be sorted by their revision.
+
+3) Commit all updates on target
+
+   All fetched updates are committed on the target. The target **must not**
+   store a revision if the corresponding update did not succeed.
+
+   This is usually possible to implement using transactions. Another way is to
+   store the revision directly in the received update (document). This way, if
+   the document update is lost, the revision will also be "lost".
```
Comment (gggeek) [inline doc/specifications/proposed/replicator/architecture.rst]: A problem:

- with legacy content engine, not all content-editing actions are atomic. E.g. change section of a node or hiding it do affect its children. So if you sync events 1) move child Y of node X away, 2) hide node X, but the sync of event 1 fails for network reasons, and you succeed to sync event 2, the final status is a discrepancy: node Y will be hidden vs. unhidden

- some actions are even harder to replicate over from master to slave in case any single update is lost (e.g. create node X then its child Y, but when syncing the creation of X is lost)

File: doc/specifications/proposed/replicator/architecture.rst [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+
+2) The updates 3, 4, 5, 6 & 7 are currently in transfer. Their state on the
+   target system is unclear until the next request for updates.
+
+3) The updates 8 & 9 are pending and not yet transferred.
+
+Given this information one could display the use the replication state of each
+document (green, yellow, red). The only additional information to store is last
+requested revision and the last transferred revision resulting from the
+requested limit.
```
Comment (gggeek) [inline doc/specifications/proposed/replicator/architecture.rst]: I think the document should go in greater detail to describe debugging and verification protocols.

In my own experience with ezcontentstaging, which has been used in real-life for quite a while, out-of-sync situations do happen, despite the best efforts in coding, and having good troubleshooting/fixing tools is a must.

**Comment by gggeek** [comment]
PS: I think that it would be valuable to get more feedback from ezcontentstaging / ezcontentsync (especially the latter) into this specification.

1. describe better the desired scenarios, and the unsupported ones.
Are all content operations supposed to be replicated, or are some  forbidden? How is the system seeded? Shall the system be able to recover automatically from out-of-sync situations? What about editors picking by hand sets of change to replicate? What about an a -> b -> c replication chain? What about replication of partial content subtrees? What if only some fields of the master data have to be pushed to the slaves, e.g. for security purposes? What if content classes or any other 'dictionary' data is out of sync? How to cope with changes over time of said dictionary data?

2. handling of ids/remote-ids to identify the data to be synchronized is not detailed

3. the problem with atomic updates / idempotent requests which is at the core of the protocol might have a deep impact on the content API. For sure, the legacy engine has unatomic operations, but I suspect that this is true for the new one as well.

**Comment by kore** [comment]
@gggeek: This document, for now, describes the general approach of handling eventual consistent replication, not detailed for the concrete use case of eZ Publish. Thanks for pointing out some of the issues, we will have to handle when going into more detail and the technical implementation.

The point will not be to create a full copy of the entire repository and all its operations. The point is to synchronize the *Content* to other indexes (Search) – this usually happens on a "view" of the repository.

Generally:

* Most likely only the one published version of Content objects will be indexed. 

* The publish() operation can probably be considered atomic (enough).

There are some very valid points, which should be considered when writing the implementation specification:

* Maintain an index of deleted documents (to know what should be deleted in the target)

* Provide information on how to debug this stack. The stack will probably be trivial, though.

* Evaluate how Compaction can work

**Comment by gggeek** [comment]
@kore I see your point. Still, it seems that any search engine used to index eZP content ends up having to deal with locations as well as contents - and then the whole floodgate of non-atomic changes comes in :-(

**Comment by andrerom** [comment]
:red_circle: :flags: Warning: there is no intention of recreating ezcontentstaging, there is plans to handle "staging" separately from any kind of replication.

This is mainly a proposal from @kore to propose how we should deal with Search engines when they get out of sync, hence why it is generic. Current approach for the foreseeable future on search engines going out of sync, is that people need to reindex content, just like in legacy.

So in summary, great input as always @gggeek, but not all of them belongs here.

**Comment by gggeek** [comment]
It's definitely not my responsibility to decide the roadmap for the eZ platform, so take this as mere advice.
But imho it would be a bit of a shame to design a system which introduces the concept of replicating changesets without taking into account all of the goodies which might come along with it (full/partial content replication)

Nice emoji btw :-)

**Comment by andrerom** [comment]
I fully agree it is an technical possibility/opportunity. But in light of the numerous attempts to tackle such an future in the past, we won't do so lightly ;)
